### PR TITLE
l2l3_config_invalid_params

### PR DESCRIPTION
**Pull request by sjwahub**
Signed-off-by: Hiren Thakkar <hiren.thakkar@intel.com>

This PR is to cover l2l3 config negative automation cases.

Test logs are:
1. Rebuild OVS log  [RebuildOVS.txt](https://github.com/ipdk-io/ovs/files/9440806/RebuildOVS.txt)
2. Unite test in format of running test suite.   [ptf_tests_run25.log](https://github.com/ipdk-io/ovs/files/9440812/ptf_tests_run25.log)
3.  Suite test log  [ptf_tests_runAll20.log](https://github.com/ipdk-io/ovs/files/9440813/ptf_tests_runAll20.log)

### CODE REVIEW

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-1/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/e2f993a4ad78ee33455cf0c0726d234133967fb1

**Comment by ipdksw** [comment]
Result:FAILURE JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-2/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/e2f993a4ad78ee33455cf0c0726d234133967fb1

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-3/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/e2f993a4ad78ee33455cf0c0726d234133967fb1

File: Makefile.am [hunk 1]
```diff
@@ -194,6 +194,8 @@ EXTRA_DIST = \
 	ptf_tests/common/config/lnt_2vm_vxlan_2nsvm_5min_ping.json \
 	ptf_tests/common/config/lnt_1hotplugvm_vxlan_1nsvm.json \
 	ptf_tests/common/config/config_control_port_with_link.json \
+    ptf_tests/common/config/l2l3_unsupport_tbl_port_match_action.json \
```
Comment (5abeel) [inline Makefile.am]: Alignment needs to be fixed

File: Makefile.am [hunk 2]
```diff
@@ -265,6 +267,8 @@ EXTRA_DIST = \
 	ptf_tests/tests/lnt_2vm_vxlan_2nsvm_5min_ping.py \
 	ptf_tests/tests/lnt_1hotplugvm_vxlan_1nsvm.py \
 	ptf_tests/tests/config_control_port_with_link.py \
+    ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py \
```
Comment (5abeel) [inline Makefile.am]: Alignment needs to be fixed

File: ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2022 Intel Corporation.
+#
```
Comment (5abeel) [inline ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py:3]: Please add the following line after the Copyright line

`# SPDX-License-Identifier: Apache-2.0`

For future reference, please add the SPDX line for all new files added.

File: ptf_tests/tests/l2l3_unsupport_tbl_port_proto_action.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
+# Copyright (c) 2022 Intel Corporation.
+#
```
Comment (5abeel) [inline ptf_tests/tests/l2l3_unsupport_tbl_port_proto_action.py:3]: Same as above. Add SPDX license line

**Review by 5abeel — CHANGES_REQUESTED** [review]

File: ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    print (f"EXPECTED failure to add Invalid match action {match_action}")
+            elif table['name'] == "ingress.ipv4_host_dst":
+                for match_action in table['match_action']:
+                    if not ovs_p4ctl.ovs_p4ctl_add_entry(table['switch'],table['name'], match_action):
+                        self.result.addFailure(self, sys.exc_info())
+                        self.fail(f"Failed to add table entry {match_action}")
+
+        print (f"Sending Traffic")
+        # Verify support table, ports and match actions
+        for table in self.config_data['table'][1:2]:
```
Comment (dkaranjai) [inline ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py]: the json file shows 2 entries for "table", unsupported and supported. the slicing here [1:2] picks the first one which is "unsupported". But for this we could have simply written as self.config['table'][0]. Why are we using slicing here?

File: ptf_tests/tests/l2l3_unsupport_tbl_port_proto_action.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                        self.fail(f"Failed: invalid entry {match_action} should not be added")       
+            elif table['name'] == "ingress.ipv4_host_dst":
+                for match_action in table['match_action']:
+                    if not ovs_p4ctl.ovs_p4ctl_add_entry(table['switch'],table['name'], match_action):
+                        self.result.addFailure(self, sys.exc_info())
+                        self.fail(f"Failed to add table entry {match_action}")
+
+        print (f"Sending Traffic")
+        # Verify support table, ports, proto, actions
+        for table in self.config_data['table'][2:3]:
```
Comment (dkaranjai) [inline ptf_tests/tests/l2l3_unsupport_tbl_port_proto_action.py]: same as above, role of slicing is not clear. you might want to add a comment block here for better readability.

**Review by dkaranjai — CHANGES_REQUESTED** [review]

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-4/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/5007819b428aaec2d5dab8958d47b2cee46a808d

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-5/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/5007819b428aaec2d5dab8958d47b2cee46a808d

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-6/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/5007819b428aaec2d5dab8958d47b2cee46a808d

File: ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py (hunk 5 above)
Comment (sjwahub) [reply ptf_tests/tests/l2l3_unsupport_tbl_port_match_action.py]: I thought a list would be extensible and used [1:2] to get a sliced list. Anyway, the code is changed to

       _for i in range(len(self.config_data['table'][1]['match_action'])):_

For this code change, unit test provided here [ptf_tests_runAll4.log](https://github.com/ipdk-io/ovs/files/9447861/ptf_tests_runAll4.log)

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-7/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/c8a68be511f0dead729087ae2ad1a0b3c1a561be

**Review by dkaranjai — APPROVED** [review]

**Review by 5abeel — APPROVED** [review]
lgtm

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-8/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/2945fec883c88c4dc9c04c7f330fc8b320cad39c

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-9/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b1eb4ef04b95e768ff25f526f4afd887baa5a3d5

**Comment by ipdksw** [comment]
Result:ABORTED JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-10/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b1eb4ef04b95e768ff25f526f4afd887baa5a3d5

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-11/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b1eb4ef04b95e768ff25f526f4afd887baa5a3d5

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-12/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b1eb4ef04b95e768ff25f526f4afd887baa5a3d5

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-13/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/b1eb4ef04b95e768ff25f526f4afd887baa5a3d5

**Comment by ipdksw** [comment]
Result:ABORTED JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-14/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/13f6d35859bedd8eb58254f1f1277cc14baeae74

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-15/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/13f6d35859bedd8eb58254f1f1277cc14baeae74

**Comment by ipdksw** [comment]
Result:ABORTED JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-16/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/13f6d35859bedd8eb58254f1f1277cc14baeae74

**Comment by ipdksw** [comment]
Result:SUCCESS JenkinsConsole:https://s3.ipdk.io/ovs/PR-61-18/index.html GitCommit:https://github.com/ipdk-io/ovs/commits/d22819ef0fd6f673e1df4cabfe19d079ede2ea1e

**Pull request merged**